src/v/cluster/partition_balancer_planner.cc

```diff
@@ -358,6 +358,105 @@ void partition_balancer_planner::get_unavailable_nodes_reassignments(
     }
 }
 
+/// Try to fix ntps that have several replicas in one rack (these ntps can
+/// appear because rack awareness constraint is not a hard constraint, e.g. when
+/// a rack dies and we move all replicas that resided on dead nodes to live
+/// ones).
+///
+/// We go over all such ntps (a list maintained by partition_balancer_state) and
+/// if the number of currently live racks is more than the number of racks that
+/// the ntp is replicated to, we try to schedule a move. For each rack we
+/// arbitrarily choose the first appearing replica to remain there (note: this
+/// is probably not optimal choice).
+void partition_balancer_planner::get_rack_constraint_repair_reassignments(
```

Comment:
Maybe we should add `rack_constraint` violations into `partition_balancer_violations`?

Reply:
I thought about it but decided against it - a full list of partitions in the violations doesn't make sense (could be thousands), but for the number of violations it makes more sense to have it as a metric - easier to observe and alert on.